### PR TITLE
Split build can now produce multiple deploy artifacts - on a per-offset basis.

### PR DESCRIPTION
**Pull request by nyarly**

### CODE REVIEW

**Comment by sous-machine-account** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR: 1955 lines of code changed</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: ext/docker/builder.go [hunk 1]
```diff
@@ -51,51 +51,56 @@ func NewBuilder(nc sous.Inserter, drh string, sourceShell, scratchShell shell.Sh
...
+		err = b.recordName(prod)
+		if err != nil {
+			return err
+		}
 	}
+	return nil
+}
 
-	return b.recordName(br, bc)
+func (b *Builder) debug(msg string) {
```
Comment (adamzr) [inline ext/docker/builder.go]: Why is this needed?

File: lib/buildpack.go [hunk 2]
```diff
@@ -67,14 +67,29 @@ type (
...
-		Elapsed                   time.Duration
-		ExtraResults              map[string]*BuildResult
+		Elapsed  time.Duration
+		Products []*BuildProduct
 	}
 
-	// EchoSelector wraps a buildpack Factory. But why?
+	// A BuildProduct is one of the individual outputs of a buildpack.
+	BuildProduct struct {
+		// Source and Kind identify the build - the source inputs and the kind of build product
```
Comment (adamzr) [inline lib/buildpack.go]: Move the documentation for Kind to its own line above Kind per golang convention

Comment (adamzr) [inline lib/buildpack.go]: The same for the other fields below.

File: dev_support/sous_qa_setup/desc/desc.go [hunk 3]
```diff
@@ -56,7 +56,7 @@ func LoadDesc(descPath string) (EnvDesc, error) {
 	dec := json.NewDecoder(descReader)
 	err = dec.Decode(&desc)
 
-	return desc, nil
+	return desc, err
```
Comment (adamzr) [inline dev_support/sous_qa_setup/desc/desc.go]: Consider using named returns.

File: ext/docker/builder.go [hunk 4]
```diff
@@ -51,51 +51,56 @@ func NewBuilder(nc sous.Inserter, drh string, sourceShell, scratchShell shell.Sh
...
-func (b *Builder) ApplyMetadata(br *sous.BuildResult, bc *sous.BuildContext) error {
-	return b.applyMetadata(br, "", bc)
+func (b *Builder) info(msg string) {
+	Log.Info.Printf(msg)
 }
 
-func (b *Builder) applyMetadata(br *sous.BuildResult, kind string, bc *sous.BuildContext) error {
-	br.VersionName = b.VersionTag(bc.Version(), kind)
-	br.RevisionName = b.RevisionTag(bc.Version(), kind, time.Now())
+func (b *Builder) applyMetadata(bp *sous.BuildProduct) error {
```
Comment (adamzr) [inline ext/docker/builder.go]: I generally prefer more descriptive names, i.e. `buildProduct` vs `bp`, but I'm not familiar with go conventions.

File: ext/docker/dockerfile_buildpack.go [hunk 5]
```diff
@@ -49,8 +51,31 @@ func NewDockerfileBuildpack() *DockerfileBuildpack {
...
+func (d *DockerfileBuildpack) Detect(c *sous.BuildContext) (*sous.DetectResult, error) {
+	dfPath := filepath.Join(c.Source.OffsetDir, "Dockerfile")
+	if !c.Sh.Exists(dfPath) {
+		return nil, fmt.Errorf("%s does not exist", dfPath)
+	}
+	sh := c.Sh.Clone()
+	sh.LongRunning(false)
+	df, err := sh.Stdout("cat", dfPath)
+	if err != nil {
+		return nil, err
```
Comment (adamzr) [inline ext/docker/dockerfile_buildpack.go]: Maybe we should wrap the error to put it in context, i.e. "The attempt to read the Dockerfile failed with the error message: ...".

File: ext/docker/dockerfile_buildpack.go [hunk 6]
```diff
@@ -12,7 +12,9 @@ import (
 
 // DockerfileBuildpack is a simple buildpack for building projects using
 // their own Dockerfile.
-type DockerfileBuildpack struct{}
+type DockerfileBuildpack struct {
+	detected *sous.DetectResult
```
Comment (adamzr) [inline ext/docker/dockerfile_buildpack.go]: Add godoc descriptive comment

File: ext/docker/dockerfile_buildpack.go [hunk 7]
```diff
@@ -49,8 +51,31 @@ func NewDockerfileBuildpack() *DockerfileBuildpack {
...
+		HasAppVersionArg:  hasAppVersion,
+		HasAppRevisionArg: hasAppRevision,
+	}}
+	return result, nil
+}
+
 // Build implements Buildpack.Build
-func (d *DockerfileBuildpack) Build(c *sous.BuildContext, dr *sous.DetectResult) (*sous.BuildResult, error) {
+func (d *DockerfileBuildpack) Build(c *sous.BuildContext) (*sous.BuildResult, error) {
+	dr := d.detected
```
Comment (adamzr) [inline ext/docker/dockerfile_buildpack.go]: `dr`?

File: ext/docker/splitcontainer_buildpack.go [hunk 8]
```diff
@@ -187,229 +103,20 @@ func (sbp *SplitBuildpack) Build(ctx *sous.BuildContext, drez *sous.DetectResult
...
-		Advisories: ctx.Advisories,
-		ExtraResults: map[string]*sous.BuildResult{
-			"builder": {
-				ImageID:    script.buildImageID,
-				Elapsed:    time.Since(start),
-				Advisories: []string{string(sous.IsBuilder)},
-			},
+		Elapsed: time.Since(start),
+		Products: []*sous.BuildProduct{
+			//{ID: script.deployImageID},
```
Comment (adamzr) [inline ext/docker/splitcontainer_buildpack.go]: Can this line be removed?

File: ext/docker/splitcontainer_detector.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+func (sd *splitDetector) absorbDockerfile() error {
+	return sd.absorbDocker(sd.rootAst)
+}
+
+func (sd *splitDetector) fetchFromRunSpec() error {
+	for _, f := range sd.froms {
+		logging.Log.Debug.Printf("Fetching FROM %q...", f.Value)
+		md, err := sd.registry.GetImageMetadata(f.Value, "")
+		if err != nil {
+			logging.Log.Debug.Printf("Error fetching %q: %v.", f.Value, err)
```
Comment (adamzr) [inline ext/docker/splitcontainer_detector.go]: Why only at the debug level?

File: ext/docker/splitcontainer_detector.go [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+			continue
```
Comment (adamzr) [inline ext/docker/splitcontainer_detector.go]: Why continue? Shouldn't we stop processing?

File: ext/docker/splitcontainer_detector.go [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+		if path, ok := md.Env[SOUS_RUN_IMAGE_SPEC]; ok {
+			logging.Log.Debug.Printf("RunSpec path %q found in %q", path, f.Value)
+			sd.runspecPath = path
+		}
+
+		buf := bytes.NewBufferString(strings.Join(md.OnBuild, "\n"))
+		ast, err := parseDocker(buf)
+		logging.Log.Debug.Printf("Parsing ONBUILD from %q.", f.Value)
+		if err != nil {
+			logging.Log.Debug.Printf("Error while parsing ONBUILD from %q: %#v.", f.Value, err)
```
Comment (adamzr) [inline ext/docker/splitcontainer_detector.go]: Why at the debug level? Should we wrap the error?

File: ext/docker/splitcontainer_splitbuilder.go [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+		sb.subBuilders = append(sb.subBuilders, &runnableBuilder{
+			RunSpec:      spec,
+			splitBuilder: sb,
+		})
+	}
+
+	return nil
+}
+
+func (sb *splitBuilder) eachBuilder(f func(*runnableBuilder) error) error {
```
Comment (adamzr) [inline ext/docker/splitcontainer_splitbuilder.go]: Shouldn't the function name be a verb?

File: test/http_state_manager_test.go [hunk 13]
```diff
@@ -72,14 +72,6 @@ func TestWriteState(t *testing.T) {
 
 	di := graph.BuildBaseGraph(&bytes.Buffer{}, os.Stderr, os.Stderr)
 	graph.AddNetwork(di)
-	/*
```
Comment (adamzr) [inline test/http_state_manager_test.go]: Can this be removed?

**Comment by nyarly** [comment]
@samsalisbury If you have bandwidth to review this, let me know. Otherwise Adam's been good enough to review.

File: .gitignore [hunk 14]
```diff
@@ -5,7 +5,8 @@ sous
 !sous/
 .agignore
 check-compile.results
-testdata/
+**/testdata/gen/
+.git/
```
Comment (nyarly) [inline .gitignore]: There are git repos as test data which we don't want checked in.

File: ext/docker/builder.go (hunk 1 above)
Comment (nyarly) [inline ext/docker/builder.go]: We're still experimenting with the best way to manage logging. I think we have a path, and once we do, we'll standardize everywhere. In the meantime, it makes debug logging less verbose.

File: ext/docker/builder.go (hunk 4 above)
Comment (nyarly) [inline ext/docker/builder.go]: Go's tradeoff (which I've come around to from `long_descriptive_ruby_snake_case_names`) is that the length of the name should be relative to the distance it covers - so struct fields have spelledOutNames generally, but the parameters of short functions or quick temporary variables are a few characters - in Java I've seen things like `catch (NullPointerException npe) ...` which is essentially a special case of this reasoning.

File: lib/buildpack.go (hunk 2 above)
Comment (nyarly) [inline lib/buildpack.go]: Fair.

File: dev_support/sous_qa_setup/desc/desc.go (hunk 3 above)
Comment (nyarly) [inline dev_support/sous_qa_setup/desc/desc.go]: As discussed, we've maybe overcorrected for using named returns too much. But c.f. https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters (which Sam linked to more than once :wink:)

File: ext/docker/dockerfile_buildpack.go (hunk 6 above)
Comment (nyarly) [inline ext/docker/dockerfile_buildpack.go]: `detected` won't be exported, so while a comment might be useful, it wouldn't be godoc'd.

File: ext/docker/dockerfile_buildpack.go (hunk 7 above)
Comment (nyarly) [inline ext/docker/dockerfile_buildpack.go]: "detect result" - short because it's an intermediate variable to avoid reusing `d.detected`

File: ext/docker/dockerfile_buildpack.go (hunk 5 above)
Comment (nyarly) [inline ext/docker/dockerfile_buildpack.go]: This is a good idea. This code dates from around when we discovered Dave Cheney's errors package, which makes that kind of contextualization (including a stacktrace) simple to do.

File: ext/docker/splitcontainer_buildpack.go (hunk 8 above)
Comment (nyarly) [inline ext/docker/splitcontainer_buildpack.go]: Yes.

File: ext/docker/splitcontainer_detector.go (hunk 10 above)
Comment (nyarly) [inline ext/docker/splitcontainer_detector.go]: It's possible that our parse of the Dockerfile is incorrect, or that there's some reason that it would be okay for the FROM line not to fetch? If it is broken, the `docker build` will catch it (and is the definitive test) - if not, we interfere with a good build because there's a parsing glitch. That, at least, was my reasoning.

File: ext/docker/splitcontainer_splitbuilder.go (hunk 12 above)
Comment (nyarly) [inline ext/docker/splitcontainer_splitbuilder.go]: I guess it could be extended to `forEachBulderDo...` but the verb would be the func `f`. Again, I'm influenced by Ruby, where it'd be simply `each`

File: test/http_state_manager_test.go (hunk 13 above)
Comment (nyarly) [inline test/http_state_manager_test.go]: Almost certainly.
